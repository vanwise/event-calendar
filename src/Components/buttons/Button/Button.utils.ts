import { css } from 'styled-components/macro';

export const BUTTON_THEMES = {
  violet: css`
    color: white;

    &::before {
      background: var(--violet);
    }

    &::after {
      border: 1px solid var(--violet);
      background: var(--red);
    }
  `,
  light: css`
    color: var(--violet);

    &::before {
      background: var(--violet2);
    }

    &::after {
      background: var(--red2);
    }
  `,
};

export const BUTTON_LOADER_COLOR = {
  violet: 'white',
  light: 'var(--violet)',
};
