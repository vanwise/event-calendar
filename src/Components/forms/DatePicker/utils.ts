import { Dispatch, SetStateAction } from 'react';
import { Dayjs } from 'dayjs';
import TimeService, { TimeServiceDate } from 'Services/TimeService';
import { DayCell } from './components/Calendar/Calendar';
import {
  SelectedDates,
  SelectedDate,
  SelectedDateType,
  DefaultSelectedDates,
} from './DatePicker';

export function getInitialSelectedDates({ from, to }: DefaultSelectedDates) {
  const dateFrom = from ? TimeService.getDate(from) : null;
  const dateTo = to ? TimeService.getDate(to) : null;

  return {
    from: dateFrom?.isValid() ? dateFrom : null,
    to: dateTo?.isValid() ? dateTo : null,
  };
}

export function getDaysInMonth(date: Dayjs, selectedDates: SelectedDates) {
  const daysInCurrentMonth = date.daysInMonth();
  let days: DayCell[] = [...Array(daysInCurrentMonth)].map((_, index) => ({
    date: date.date(index + 1),
    isActive: true,
  }));

  addMarkOfDatesFromAndToInDays();
  setPrevMonthDays();
  setNextMonthDays();

  function addMarkOfDatesFromAndToInDays() {
    const isMonthSameSelectedDatesMonth = Boolean(
      selectedDates.from?.isSame(date, 'month') ||
        selectedDates.to?.isSame(date, 'month'),
    );

    if (isMonthSameSelectedDatesMonth) {
      days = days.map(day => ({
        ...day,
        isFrom: selectedDates.from?.isSame(day.date, 'date'),
        isTo: selectedDates.to?.isSame(day.date, 'date'),
      }));
    }
  }

  function setPrevMonthDays() {
    const weekDayOfStartOfMonth = date.startOf('month').weekday();
    const prevMonth = date.subtract(1, 'month');
    const daysInPrevMonth = prevMonth.daysInMonth();

    [...Array(weekDayOfStartOfMonth)].forEach((_, index) => {
      days.unshift({
        date: prevMonth.date(daysInPrevMonth - index),
      });
    });
  }

  function setNextMonthDays() {
    const weekDayOfEndOfMonth = date.endOf('month').weekday();
    const nexMonth = date.add(1, 'month');

    [...Array(6 - weekDayOfEndOfMonth)].forEach((_, index) => {
      days.push({
        date: nexMonth.date(index + 1),
      });
    });
  }

  return days;
}

interface HandleCalendarDateClickArgs {
  newDate: TimeServiceDate;
  currentDate: TimeServiceDate;
  selectedDates: SelectedDates;
  setCurrentDate: Dispatch<SetStateAction<TimeServiceDate>>;
  updateSelectedDate(type: SelectedDateType, newDate: SelectedDate): void;
  deleteSelectedDate(type: SelectedDateType): void;
  updateSelectedDates(newDates: SelectedDates): void;
}

export function handleCalendarDateClick({
  newDate,
  currentDate,
  selectedDates,
  setCurrentDate,
  updateSelectedDate,
  deleteSelectedDate,
  updateSelectedDates,
}: HandleCalendarDateClickArgs) {
  if (newDate.isSame(currentDate, 'month')) {
    toggleSelectedDate();
  } else {
    switchToOtherMonth();
  }

  function toggleSelectedDate() {
    if (selectedDates.from?.isSame(newDate, 'date')) {
      deleteSelectedDate('from');
    } else if (selectedDates.to?.isSame(newDate, 'date')) {
      deleteSelectedDate('to');
    } else if (selectedDates.from && selectedDates.to) {
      setCorrectUpdatedDate(selectedDates.from, selectedDates.to);
    } else if (selectedDates.from || selectedDates.to) {
      setCorrectNewDate();
    } else {
      updateSelectedDate('from', newDate);
    }

    function setCorrectUpdatedDate(dateFrom: Dayjs, dateTo: Dayjs) {
      const middleDateBetweenSelectedDates = dateFrom.add(
        dateTo.diff(dateFrom, 'millisecond') / 2,
        'millisecond',
      );
      const isDateBeforeMiddleDate = newDate.isBefore(
        middleDateBetweenSelectedDates,
        'date',
      );

      if (isDateBeforeMiddleDate) {
        updateSelectedDate('from', newDate);
      } else {
        updateSelectedDate('to', newDate);
      }
    }

    function setCorrectNewDate() {
      const isDateAfterDateTo = newDate.isAfter(selectedDates.to, 'date');
      const isDateBeforeDateFrom = newDate.isBefore(selectedDates.from, 'date');

      if (!selectedDates.to && isDateBeforeDateFrom) {
        updateSelectedDates({
          to: selectedDates.from,
          from: newDate,
        });
      } else if (!selectedDates.from && isDateAfterDateTo) {
        updateSelectedDates({
          to: newDate,
          from: selectedDates.to,
        });
      } else if (selectedDates.to) {
        updateSelectedDate('from', newDate);
      } else if (selectedDates.from) {
        updateSelectedDate('to', newDate);
      }
    }
  }

  function switchToOtherMonth() {
    if (newDate.isBefore(currentDate, 'month')) {
      setCurrentDate(currentDate.subtract(1, 'month'));
    } else if (newDate.isAfter(currentDate, 'month')) {
      setCurrentDate(currentDate.add(1, 'month'));
    }
  }
}
