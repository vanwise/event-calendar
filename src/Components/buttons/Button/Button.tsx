import styled from 'styled-components/macro';
import Loader from 'Components/Loader';
import { BUTTON_LOADER_COLOR, BUTTON_THEMES } from './Button.utils';

type Theme = keyof typeof BUTTON_THEMES;
interface ButtonProps extends ButtonPropsWithoutRef {
  theme?: Theme;
  isLoading?: boolean;
}

function Button({
  theme = 'violet',
  children,
  disabled,
  isLoading,
  ...props
}: ButtonProps) {
  const loaderColor = BUTTON_LOADER_COLOR[theme];

  return (
    <Root
      type="button"
      $theme={theme}
      disabled={isLoading || disabled}
      {...props}>
      {isLoading ? (
        <Loader color={loaderColor} width={4} height={20} />
      ) : (
        children
      )}
    </Root>
  );
}

const Root = styled.button<{ $theme: Theme }>`
  position: relative;
  z-index: 0;
  padding: 15px;
  min-height: 55px;
  transition: 0.3s ease-out;
  border-radius: 10px;
  font-size: 18px;

  &::before {
    content: '';
    position: absolute;
    z-index: -1;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border-radius: 10px;
    transition: 0.3s ease-out;
  }

  &::after {
    content: '';
    position: absolute;
    z-index: -2;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border-radius: 10px;
    transition: 0.3s ease-out;
  }

  &:hover:not([disabled]) {
    transform: translate(-3px, -3px);

    &::after {
      transform: translate(5px, 5px);
    }
  }

  ${({ $theme }) => BUTTON_THEMES[$theme]}
`;

export default Button;
