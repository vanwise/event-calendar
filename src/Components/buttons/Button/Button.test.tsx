import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button, { ButtonTheme } from './Button';
import { BUTTON_THEMES } from './Button.utils';

const onClick = jest.fn();

function setup(props = {}) {
  const renderResult = render(
    <Button onClick={onClick} {...props}>
      Test
    </Button>,
  );
  const button = renderResult.getByRole('button', { name: 'Test' });

  return [button, renderResult] as const;
}

afterEach(jest.clearAllMocks);

describe('Button', () => {
  it('should be clicked', async () => {
    const [button] = setup();

    await userEvent.click(button);
    expect(onClick).toBeCalledTimes(1);
  });

  it('should be disabled and show loader when loading status is active', async () => {
    const [button] = setup({ isLoading: true });

    const loader = await screen.findByTestId('loader');

    expect(button).toBeDisabled();
    await userEvent.click(button);
    expect(onClick).not.toBeCalled();

    expect(loader).toBeInTheDocument();
  });

  it('should have diferrent themes', () => {
    const [button, { rerender }] = setup();

    for (const theme in BUTTON_THEMES) {
      rerender(<Button theme={theme as ButtonTheme} />);
      expect(button).toMatchSnapshot();
    }
  });
});
