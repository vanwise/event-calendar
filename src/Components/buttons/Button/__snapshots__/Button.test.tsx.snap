// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should have diferrent themes 1`] = `
.c0 {
  position: relative;
  z-index: 0;
  padding: 15px;
  min-height: 55px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
  border-radius: 10px;
  font-size: 18px;
  color: white;
}

.c0::before {
  content: '';
  position: absolute;
  z-index: -1;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 10px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
}

.c0::after {
  content: '';
  position: absolute;
  z-index: -2;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 10px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
}

.c0:hover:not([disabled]) {
  -webkit-transform: translate(-3px,-3px);
  -ms-transform: translate(-3px,-3px);
  transform: translate(-3px,-3px);
}

.c0:hover:not([disabled])::after {
  -webkit-transform: translate(5px,5px);
  -ms-transform: translate(5px,5px);
  transform: translate(5px,5px);
}

.c0:disabled {
  color: var( --gray8 );
}

.c0:disabled::before {
  background: var(--gray);
}

.c0::before {
  background: var(--violet);
}

.c0::after {
  border: 1px solid var(--violet);
  background: var(--red);
}

<button
  class="c0"
  type="button"
/>
`;

exports[`Button should have diferrent themes 2`] = `
.c0 {
  position: relative;
  z-index: 0;
  padding: 15px;
  min-height: 55px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
  border-radius: 10px;
  font-size: 18px;
  color: var(--violet);
}

.c0::before {
  content: '';
  position: absolute;
  z-index: -1;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 10px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
}

.c0::after {
  content: '';
  position: absolute;
  z-index: -2;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 10px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
}

.c0:hover:not([disabled]) {
  -webkit-transform: translate(-3px,-3px);
  -ms-transform: translate(-3px,-3px);
  transform: translate(-3px,-3px);
}

.c0:hover:not([disabled])::after {
  -webkit-transform: translate(5px,5px);
  -ms-transform: translate(5px,5px);
  transform: translate(5px,5px);
}

.c0:disabled {
  color: var( --gray8 );
}

.c0:disabled::before {
  background: var(--gray);
}

.c0::before {
  background: var(--violet2);
}

.c0::after {
  background: var(--red2);
}

<button
  class="c0"
  type="button"
/>
`;
